fix the issue of text being cut in TextView with margin

keithlanding opened the pull request.

thagikura commented on flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java:
```diff
@@ -322,7 +322,7 @@ public void addView(View child, int index, ViewGroup.LayoutParams params) {
      */
     private int[] createReorderedIndices(View viewBeforeAdded, int indexForViewBeforeAdded,
-            ViewGroup.LayoutParams paramsForViewBeforeAdded) {
+                                         ViewGroup.LayoutParams paramsForViewBeforeAdded) {
```
Could you revert the unneeded format changes?

Ditto for the rest of the format changes below.

thagikura commented on flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java:
```diff
@@ -497,7 +497,7 @@ private void measureHorizontal(int widthMeasureSpec, int heightMeasureSpec) {
                         child.getMeasuredHeight() + lp.topMargin + lp.bottomMargin);
 
                 if (isWrapRequired(mFlexWrap, widthMode, widthSize, flexLine.mainSize,
```
I think you don't have to add an argument to the `isWrapRequired` method, instead you can change the fourth argument as `child.getMeasuredWidth + lp.leftMargin + lp.rightMargin`.

thagikura commented on flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java:
```diff
@@ -648,7 +648,7 @@ private void measureVertical(int widthMeasureSpec, int heightMeasureSpec) {
                     child.getMeasuredWidth() + lp.leftMargin + lp.rightMargin);
 
             if (isWrapRequired(mFlexWrap, heightMode, heightSize, flexLine.mainSize,
```
Likewise you can change the fourth argument as child.getMeasuredHeight() + lp.topMargin + lp.bottomMargin instead of adding a new argument.

thagikura commented on flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java:
```diff
@@ -1359,20 +1359,28 @@ private void setMeasuredDimensionForFlex(@FlexDirection int flexDirection, int w
      * @see #getFlexWrap()
      * @see #setFlexWrap(int)
      */
     private boolean isWrapRequired(int flexWrap, int mode, int maxSize,
```
You don't have to change this method as per my comment above.

thagikura commented:
Thank you for your contribution!
The passed value for the `isWrapRequired` method needs to be fixed indeed, but this doesn't fix the #62.
The children used as an example in the #62 don't have margins.

thagikura commented:
Also is it possible to add a test case in the FlexboxAndroidTest, the test would fail unless the change in this PR is merged?

keithlanding commented:
1. the argument "isHorizontal" I add is to support the directions requirment, u know the this layout support both Horizontal and Vertical, which takes diffrent params into calculation. Hozizontal mode need leftMargin and rightMargin, while Vertical mode need topMargin and bottomMargin,so this argument is needed.
2. As to the unneeded format, i don't notice them until i make the pull request, can u just 
 merge the useful part and ignore the unneeded part?

And this does solve the text being cutted problem.

keithlanding commented:
i have updated the code:
1. remove the unneeded changes;
2. remove the extra argument i added.
3. the code fix the issue of text being cut in TextView with margin, which is different with #62 ,sorry for the wrong title, but the bug of text being cut does exit and this will fix it.

My question:
1. Where does this unneeded changes (extra spaces) come frome? as i have done as you said to set AndroidStyle as my project style and format it.

thagikura commented on flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java:
```diff
@@ -1359,28 +1359,28 @@ private void setMeasuredDimensionForFlex(@FlexDirection int flexDirection, int w
+        // the issue of text being cut in TextView with margin comes from here, because it didn't take margins into account
         // so i add the margin into the calculation to fix the bug
-        if (isHorizontal)
+        if (mFlexDirection == FLEX_DIRECTION_ROW || mFlexDirection == FLEX_DIRECTION_ROW_REVERSE) {
```
Could you revert the isWrapRequired method?

As in my previous comment, you can add lp.leftMargin + lp.rightMargin (or lp.topMargin + lp.bottomMargi) to the fourth when isWrapRequired is called in measureHorizontal/measureVertical

thagikura commented:
Thanks for the update.
Left some comments.

> My question:
> 1. Where does this unneeded changes (extra spaces) come frome? as i have done as you said to set AndroidStyle as my project style and format it.
Please make sure you imported the correct format file.

And how about a test?
> Also is it possible to add a test case in the FlexboxAndroidTest, the test would fail unless the change in > this PR is merged?

keithlanding commented:
@thagikura  
1. Thanks for your advice, i have updated the code with adding the margins to the fifth argument of isWrapReuiqred and leave it unchanged.
2. Continue my question of the **Style Problem**, could u tell me the specific name of the "extra spaces" happen to me in the previous code? which i can go to check the style and find where is the problem.
3. As the test case, you mind give me a hint how to add it ?

thagikura commented:
Hi,

Thanks for your continuous effort.
For 2, sorry I don't understand about the question 2. The #62 should be fixed by #70. And as I commented in the code review, this PR is a different (potential) issue where there is a flex line wrapping problem in the case flex line wrapping doesn't happen (but should've happened) because of the child's horizontal margin (or vertical margin) isn't taken into account when judging if a wrapping is required.

For 3, please refer to the existing FlexboxAndroidTest under the flexbox/src/androidTest.

keithlanding commented:
@thagikura  i have updated the code with a test case

thagikura commented:
@keithlanding 

Thanks for the update.
I'm so sorry... after another consideration, I realized #70 wasn't a correct fix for #62.
And realized #74 is the right way to fix it.
The problem was when judging if a new flex line is required (isWrapRequired method), the padding values from the parent and the margin values from the child weren't taken into account.

The change in #74 overlap the change in this PR.
(Actually #62 was caused by the missing padding values from the parent, but child's margin values were also missing for the isWrapRequired method)
So I'd like to merge the #74 and close this PR.

Apologies for not being able to merge this PR. At first I thought the change in this PR didn't overlap my PR.. I appreciate if you understand the situation.

Thanks.

keithlanding commented:
@thagikura Got it, go ahead, fixing the issue is what matters.
